### PR TITLE
[Major Rework] Replace Stack with StackPointer<CanvasStack> and use Runtime features

### PR DESCRIPTION
**Pull request by jtenner**
This pull request aims to replace a lot of the old pre-runtime AssemblyScript code with appropriate revisions that are faster and comply with the runtime using some manual memory management techniques I learned while writing software for the std lib.

There is lots of improvement to make and any help would be appreciated.

### CODE REVIEW

File: assembly/draw-functions.test.ts [hunk 1]
```diff
@@ -19,7 +17,7 @@ import {
 
 var ctx: CanvasRenderingContext2D;
 var grd: CanvasGradient;
-var img: Image;
+var img: Image | null = null;
```
Comment (willemneal) [inline assembly/draw-functions.test.ts]: Why must this be null?

Comment (jtenner) [reply assembly/draw-functions.test.ts]: This value gets recreated all the time, and it was erroring whenever I tried to compile it otherwise.

**Comment by jtenner** [comment]
Looks like this needs to be updated to the npm package for testing purposes. Will have to take care of this soon.

**Comment by jtenner** [comment]
@willemneal Thank you for your help.

**Pull request merged**